kernel/work.c

```diff
@@ -0,0 +1,952 @@
+static int _cancel_async(struct k_work *work)
+{
+	/* If we haven't already started canceling, do it now. */
+	if (!atomic_test_bit(&work->flags, K_WORK_CANCELING_BIT)) {
+		/* Remove it from the queue, if it's queued. */
+		queue_remove(work->queue, work);
+	}
+
+	/* If it's still busy after it's been dequeued, then flag it
+	 * as canceling.
+	 */
+	int ret = k_work_busy(work);
+
+	if (ret != 0) {
+		atomic_set_bit(&work->flags, K_WORK_CANCELING_BIT);
+		ret = k_work_busy(work);
+	}
+
+	return ret;
+}
+
+/* Complete cancellation necessary, release work lock, and wait if
+ * necessary.
+ *
+ * Invoked with work lock held by key.
+ * Releases lock.
+ * Sleeps.
+ *
+ * @param work work that is being canceled
+ * @param canceller state used to synchronize the cancellation
+ * @param key used by work lock
+ *
+ * @return true iff the work was still active on entry (waited)
+ * @return false if work was idle on entry (no wait required)
+ */
+static bool _cancel_sync_and_unlock(struct k_work *work,
+				    struct z_work_canceller *canceller,
+				    k_spinlock_key_t key)
+{
+	bool ret = false;
+
+	/* If something's still running then we have to wait for
+	 * completion, which is indicated when finish_cancel() gets
+	 * invoked.
+	 */
+	if (atomic_test_bit(&work->flags, K_WORK_CANCELING_BIT)) {
+		init_work_cancel(canceller, work);
+		ret = true;
+
+		k_spin_unlock(&lock, key);
+
+		k_sem_take(&canceller->sem, K_FOREVER);
+	} else {
+		k_spin_unlock(&lock, key);
+	}
+
+	return ret;
+}
+
+int k_work_cancel(struct k_work *work)
+{
+	__ASSERT_NO_MSG(work != NULL);
+	__ASSERT_NO_MSG(!atomic_test_bit(&work->flags, K_WORK_DELAYABLE_BIT));
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	int ret = _cancel_async(work);
+
+	k_spin_unlock(&lock, key);
+
+	return ret;
+}
+
+bool k_work_cancel_sync(struct k_work *work,
+			struct k_work_sync *sync)
+{
+	__ASSERT_NO_MSG(work != NULL);
+	__ASSERT_NO_MSG(sync != NULL);
+	__ASSERT_NO_MSG(!atomic_test_bit(&work->flags, K_WORK_DELAYABLE_BIT));
+	__ASSERT_NO_MSG(!k_is_in_isr());
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	(void)_cancel_async(work);
+
+	return _cancel_sync_and_unlock(work, &sync->canceller, key) != 0;
+}
+
+/* Work has been dequeued and is about to be invoked by the work
+ * thread.
+ *
+ * If the work is being canceled the cancellation will be completed
+ * here, and the caller told not to use the work item.
+ *
+ * Invoked by work queue thread.
+ * Takes and releases lock.
+ * Reschedules via _finalize_cancel_and_unlock
+ *
+ * @param work work that is changing state
+ * @param queue queue that is running work
+ *
+ * @retval true if work is to be run by the work thread
+ * @retval false if it has been canceled and should not be run
+ */
+static inline bool work_set_running(struct k_work *work,
+				    struct k_work_q *queue)
+{
+	bool ret = false;
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	/* Allow the work to be queued again. */
+	atomic_clear_bit(&work->flags, K_WORK_QUEUED_BIT);
```

Comment:
some detail on the concurrency scenarios that use of these atomic operations are trying to mitigate would be helpful. Doesn't the global `lock` already protect concurrent access to the work items?

I'm not seeing the usual compare-and-swap design pattern with these atomic variables...seems like things could still get messed up if these atomic variables are modified outside of the spinlock

Reply:
>  so if the flags aren't atomic objects a thread on another processor could grab the lock but see an outdated copy of the flags.

Are you absolutely sure about this? That doesn't sound right.
The spinlocks themselves are barriers, the compiler shouldn't be reordering stuff around the locking calls.
If they were, an astonishing amount of things would start breaking...

Reply:
No, I'm not absolutely sure.  I went this route after running into #30360 which mixes race conditions on non-atomic data with atomic operations.  It may be the "barrier" in `k_spin_unlock()` is sufficient to avoid this, but I haven't observed similar problems since the switch to using atomic flags so I declared victory and moved on.

Reply:
One reason to keep the flags atomic: there is existing API like `k_work_busy()` which reads the flags without locking.  Somebody might use those in a loop that would be optimized away unless the atomic API is used to fetch (this is noted in the sequential consistency section of https://gcc.gnu.org/wiki/Atomic/GCCMM/AtomicSync).  I've updated the documentation to use that justification.

Reply:
I really think you are barking up the wrong tree here. The locks themselves are barriers.

> which reads the flags without locking.

This is a mistake. If data is protected by a lock, you need to use the lock for all access to it. not sometimes use it and sometimes do not.

Reply:
But it's atomic.  Reading it doesn't require a lock, just to not be optimized away, just like the current implementation does.  The API that reads it is *defined* as racy.

Reply:
Alright, unless other people chime in I'll make the flags fields part of an unsigned int, and will pull the inline state checking functions out to real functions so they can take the lock before reading the state.